staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go

```diff
@@ -17,330 +17,239 @@ limitations under the License.
 type SpecAggregator interface {
-	AddUpdateAPIService(handler http.Handler, apiService *v1.APIService) error
-	UpdateAPIServiceSpec(apiServiceName string, spec *spec.Swagger, etag string) error
-	RemoveAPIServiceSpec(apiServiceName string) error
-	GetAPIServiceInfo(apiServiceName string) (handler http.Handler, etag string, exists bool)
-	GetAPIServiceNames() []string
+	AddUpdateAPIService(apiService *v1.APIService, handler http.Handler) error
+	UpdateAPIServiceSpec(apiServiceName string) error
+	RemoveAPIService(apiServiceName string) error
 }
 
 const (
 	aggregatorUser                = "system:aggregator"
-	specDownloadTimeout           = 60 * time.Second
-	localDelegateChainNamePrefix  = "k8s_internal_local_delegation_chain_"
-	localDelegateChainNamePattern = localDelegateChainNamePrefix + "%010d"
+	specDownloadTimeout           = time.Minute
+	localDelegateChainNamePattern = "k8s_internal_local_delegation_chain_%010d"
 
 	// A randomly generated UUID to differentiate local and remote eTags.
 	locallyGeneratedEtagPrefix = "\"6E8F849B434D4B98A569B9D7718876E9-"
 )
 
-// IsLocalAPIService returns true for local specs from delegates.
-func IsLocalAPIService(apiServiceName string) bool {
-	return strings.HasPrefix(apiServiceName, localDelegateChainNamePrefix)
+// openAPISpecInfo is used to store OpenAPI specs.
+// The apiService object is used to sort specs with their priorities.
+type openAPISpecInfo struct {
+	apiService v1.APIService
+	// spec is the cached OpenAPI spec
+	spec cached.Replaceable[*spec.Swagger]
+
+	// The downloader is used only for non-local apiservices to
+	// re-update the spec every so often.
+	downloader cached.Data[*spec.Swagger]
 }
 
-// GetAPIServiceNames returns the names of APIServices recorded in specAggregator.openAPISpecs.
-// We use this function to pass the names of local APIServices to the controller in this package,
-// so that the controller can periodically sync the OpenAPI spec from delegation API servers.
-func (s *specAggregator) GetAPIServiceNames() []string {
-	names := make([]string, 0, len(s.openAPISpecs))
-	for key := range s.openAPISpecs {
-		names = append(names, key)
+type specAggregator struct {
+	// mutex protects the specsByAPIServiceName map and its contents.
+	mutex sync.Mutex
+
+	// Map of API Services' OpenAPI specs by their name
+	specsByAPIServiceName map[string]*openAPISpecInfo
+
+	// provided for dynamic OpenAPI spec
+	openAPIVersionedService *handler.OpenAPIService
+
+	downloader *Downloader
+}
+
+func buildAndRegisterSpecAggregatorForLocalServices(downloader *Downloader, aggregatorSpec *spec.Swagger, delegationHandlers []http.Handler, pathHandler common.PathHandler) *specAggregator {
+	s := &specAggregator{
+		downloader:            downloader,
+		specsByAPIServiceName: map[string]*openAPISpecInfo{},
+	}
+	cachedAggregatorSpec := cached.NewStaticSource(func() cached.Result[*spec.Swagger] {
+		return cached.NewResultOK(aggregatorSpec, "never-changes")
+	})
```

Comment:
```suggestion
	cachedAggregatorSpec := cached.NewResultOK(aggregatorSpec, "never-changes")
```

Reply:
I feel dejavu from this. Thanks for pointing it out, fixed!